File: src/main/java/com/mapk/core/BucketGenerator.java
package com.mapk.core;

import com.mapk.annotations.KParameterRequireNonNull;
import kotlin.Pair;
import kotlin.reflect.KParameter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

class BucketGenerator {
    private final Boolean[] initializationStatus;

    @NotNull
    private final List<Boolean> isRequireNonNull;
    @NotNull
    private final KParameter[] keyArray;
    @NotNull
    private final Object[] valueArray;

    BucketGenerator(List<KParameter> parameters, @Nullable Pair<KParameter, Object> instancePair) {
        final int capacity = parameters.size();

        isRequireNonNull = new ArrayList<>(capacity);
        initializationStatus = new Boolean[capacity];

        for (int i = 0; i < capacity; i++) {
            isRequireNonNull.add(
                    i,
                    parameters.get(i).getAnnotations().stream().anyMatch(it -> it instanceof KParameterRequireNonNull)
            );
            initializationStatus[i] = false;
        }

        keyArray = new KParameter[capacity];
        valueArray = new Object[capacity];

        if (instancePair != null) {
            keyArray[0] = instancePair.getFirst();
            valueArray[0] = instancePair.getSecond();
            initializationStatus[0] = true;
        } else {
            initializationStatus[0] = false;
        }
    }

    @NotNull
    ArgumentBucket generate() {
        return new ArgumentBucket(
                keyArray.clone(),
                valueArray.clone(),
                isRequireNonNull,
                new InitializationStatusManagerImpl(initializationStatus.clone())
        );
    }
}

File: src/main/kotlin/com/mapk/core/BucketGenerator.kt
package com.mapk.core

import com.mapk.annotations.KParameterFlatten
import kotlin.reflect.KClass
import kotlin.reflect.KParameter
import kotlin.reflect.full.findAnnotation

internal class BucketGenerator(
    private val parameters: List<KParameter>,
    filteredParameters: List<KParameter>, // フィルタリングは外でもやっているため、ここでは引数として受け取る
    instance: Any?,
    parameterNameConverter: ParameterNameConverter
) {
    private val binders: List<ArgumentBinder> = filteredParameters.map { it.toArgumentBinder(parameterNameConverter) }
    private val originalValueArray: Array<Any?> = arrayOfNulls(parameters.size)
    private val originalInitializationStatus: Array<Boolean> = Array(parameters.size) { false }
    val valueParameters: List<ValueParameter<*>>

    init {
        if (instance != null) {
            originalValueArray[0] = instance
            originalInitializationStatus[0] = true
        }

        // TODO: 仮置き、これを生成するのはKFunctionForCallの方が良さげ
        valueParameters = binders.fold(ArrayList()) { acc, elm ->
            when (elm) {
                is ArgumentBinder.Value<*> -> acc.add(elm)
                is ArgumentBinder.Function -> acc.addAll(elm.requiredParameters)
            }
            acc
        }
    }

    fun generate(adaptor: ArgumentAdaptor): ArgumentBucket = ArgumentBucket(
        parameters, originalValueArray.clone(), originalInitializationStatus.clone(), binders, adaptor
    )
}

private fun KParameter.toArgumentBinder(parameterNameConverter: ParameterNameConverter): ArgumentBinder {
    val name = getAliasOrName()!!

    return findAnnotation<KParameterFlatten>()?.let { annotation ->
        // 名前の変換処理、結合が必要な場合はインスタンスを持ってきて対応する
        val converter: ParameterNameConverter = if (annotation.fieldNameToPrefix) {
            parameterNameConverter.nest(name, annotation.nameJoiner.objectInstance!!)
        } else {
            parameterNameConverter
        }

        ArgumentBinder.Function((type.classifier as KClass<*>).toKConstructor(converter), index, annotations)
    } ?: ArgumentBinder.Value(
        index,
        annotations,
        isOptional,
        parameterNameConverter.convert(name),
        type.classifier as KClass<*>
    )
}
